server/api/submit_response.go

```diff
@@ -0,0 +1,117 @@
+		api.pluginAPI.LogError("handleSubmitSurveyResponse: failed to unmarshal request body", "error", err.Error())
+		http.Error(w, "failed to read request body", http.StatusInternalServerError)
+		return
+	}
+
+	response.SurveyID = surveyID
+	response.UserID = userID
+
+	// now that we have response, we'll verify that the response matches
+	// what is expected in the active survey. For example,
+	// the number of questions and answers should match, and the submission should
+	// only be against the active survey
+
+	survey, err := api.app.GetInProgressSurvey()
+	if err != nil {
+		api.pluginAPI.LogError("handleSubmitSurveyResponse: failed to fetch in progress survey", "error", err.Error())
+		http.Error(w, "failed to fetch survey", http.StatusInternalServerError)
+		return
+	}
+
+	if err := matchSurveyAndResponse(surveyID, survey, response); err != nil {
+		api.pluginAPI.LogError("handleSubmitSurveyResponse: failed to match survey and response", "error", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if err := api.app.SaveSurveyResponse(response); err != nil {
+		api.pluginAPI.LogError("handleSubmitSurveyResponse: failed to save survey response", "error", err.Error())
+		http.Error(w, "failed to save response", http.StatusInternalServerError)
+		return
+	}
+
+	ReturnStatusOK(w)
+}
+
+func matchSurveyAndResponse(surveyID string, survey *model.Survey, response *model.SurveyResponse) error {
+	// the response should belong to the currently active survey
+	if survey.ID != surveyID {
+		return errors.New("the survey you're responding to is no longer active")
+	}
+
+	// response can't have more answers than the number of questions in the survey
+	if len(response.Response) > len(survey.SurveyQuestions.Questions) {
+		return errors.New("incorrect number of responses submitted")
+	}
+
+	linearScaleQuestionID, err := survey.GetSystemRatingQuestionID()
+	if err != nil {
+		return err
+	}
+
+	if _, ok := response.Response[linearScaleQuestionID]; !ok {
+		return errors.New("linear scale question must be answered")
+	}
+
+	// if only one response is submitted, it needs to be
+	// the answer to the linear scale question
+	if len(response.Response) == 1 {
+		// When user selects a rating and submits via the Submit button,
+		// the client passes the response type manually, and we should only verify it,
+		// not override it.
+		if response.ResponseType != model.ResponseTypeComplete {
+			response.ResponseType = model.ResponseTypePartial
+		}
+	} else {
+		// make sure answered questions belong to the survey
+		surveyQuestionIDMap := map[string]bool{}
+		for _, question := range survey.SurveyQuestions.Questions {
+			surveyQuestionIDMap[question.ID] = true
+		}
+
+		for responseQuestionID := range response.Response {
+			if _, ok := surveyQuestionIDMap[responseQuestionID]; !ok {
+				return errors.New("invalid question ID found in submitted answer")
+			}
+		}
+
+		response.ResponseType = model.ResponseTypeComplete
```

Comment:
It needs to be done at some place, since we're already checking everything here - len of responses, checks for rating question etc, this was a nice place to add this without checking everything again.

Reply:
@harshilsharma63 If we assign the response type here when we verify that the questions matched the in-progress survey, why are we also sending the `ResponseType` in the request when the user hits the `Submit` button?

Is this due to the check in lines 103-105 where we need to be able to tell when only the partial rating result was sent (without the submit button)?